# Homework 2
### Jake Pitkin
**CS 6190: Probabilistic Modeling - Spring 2018**<br>
**February 19 2018**

**1. Create the following Bayesian network to analyze the survey results. You will want to use the provided function createCPTfromData.**

<img src="img/network.png" width="855" height="611" align="center"/>

First we will create the following Bayesian network using the Behavioral Risk Factor Surveillance System
(BRFSS) survey data. The source code for the Bayesian network is located in `bayesian_network.py`.

In [1]:
from bayesian_network import *

# Read in the dataset
data = pd.read_csv("RiskFactors.csv")

# Create the Bayesian network from the dataset
income = createCPTfromData(data, ["income"])
bmi = createCPTfromData(data, ["bmi", "income", "exercise"])
exercise = createCPTfromData(data, ["exercise", "income"])
smoke = createCPTfromData(data, ["smoke", "income"])
bp = createCPTfromData(data, ["bp", "income", "exercise", "smoke"])
cholesterol = createCPTfromData(data, ["cholesterol", "income", "smoke", "exercise"])
diabetes = createCPTfromData(data, ["diabetes", "bmi"])
stroke = createCPTfromData(data, ["stroke", "bmi", "bp", "cholesterol"])
attack = createCPTfromData(data, ["attack", "bmi", "bp", "cholesterol"])
angina = createCPTfromData(data, ["angina", "bmi", "bp", "cholesterol"])
network = [income, bmi, exercise, smoke, bp, cholesterol, diabetes, stroke, attack, angina]

**1a. What is the size (in terms of the number of probabilities needed) of this network?**

The size of the network can be determined by considering the number of rows in each factor of the network.

In [2]:
rows = 0
for factor in network:
    rows += factor.shape[0]
print(rows)

504


The size of this network is 504 probabilities.

**1b. What is the total number of probabilities needed to store the full joint distribution?**

For the entire joint distribution, we take the product of the number of values each random variable can take:

$$\prod_{i=1}^{10} \ n(domain_i) = 8 * 2 * 2 * 4 * 4 * 2 * 2 * 2 * 2 * 4 = 32,768$$

The size of the full join distribution is 32,768 probabilities.

**2. For each of the four health outcomes (diabetes, stroke, heart attack, angina), answer the following by querying your network (using your infer function):**

**2a. What is the probabilitiy of the outcome if I have bad habits (smoke and don't exercise)?**

In [3]:
diabetes_marg = ["income", "bmi", "exercise", "smoke", "bp", "cholesterol", "stroke", "attack", "angina"]
diabetes_bad_habits = infer(network, diabetes_marg, ["smoke", "exercise"], [1, 2])
print(diabetes_bad_habits)

      probs  diabetes
0  0.159330         1
1  0.007881         2
2  0.812632         3
3  0.020156         4


**2a. How about if I have good habits (don't smoke and do exercise)**

In [4]:
diabetes_marg = ["income", "bmi", "exercise", "smoke", "bp", "cholesterol", "stroke", "attack", "angina"]
diabetes_good_habits = infer(network, diabetes_marg, ["smoke", "exercise"], [2, 1])
print(diabetes_good_habits)

      probs  diabetes
0  0.135062         1
1  0.007674         2
2  0.839392         3
3  0.017873         4


**2b. What is the probability of the outcome if I have poor health (high blood pressure, high cholesterol, and overweight)?**

**2b. How about if I have good health (low blood pressure, low cholesterol, and normal weight)?**